lib/FetchEnvs.js

```diff
@@ -14,210 +14,291 @@
-          resource = await krm.get(name, namespace);
-        } catch (e) {
-          log.warn(e);
-          kubeError = e.message;
-        }
+      fn = this.#genericKeyRef;
+    }
+
+    try {
+      result = await fn.call(this, conf);
+    } catch (error) {
+      log.warn(error);
+      kubeError = error.message;
+    }
+
+    return kubeError ? { error: kubeError } : result;
+  }
+
+  async #genericMapRef(conf) {
+    const apiVersion = conf.genericMapRef?.apiVersion;
+    const kind = conf.genericMapRef?.kind;
+    const name = conf.genericMapRef?.name;
+    const namespace = conf.genericMapRef?.namespace ?? this.namespace;
+    const optional = !!conf.optional;
+    const krm = await this.kubeClass.getKubeResourceMeta(apiVersion, kind, 'update');
+    let resource;
+    let kubeError;
+
+    if (krm) {
+      try {
+        resource = await krm.get(name, namespace);
+      } catch (error) {
+        log.warn(error);
+        kubeError = error.message;
       }
-      key = objectPath.get(valueFrom, 'genericKeyRef.key');
-      result = objectPath.get(resource, ['data', key]);
-      type = objectPath.get(valueFrom, 'genericKeyRef.type');
     }
-    if (type && result) {
-      switch (type) {
-        case 'number':
-          result = Number(result);
-          break;
-        case 'boolean':
-          result = (result.toLowerCase() === 'true');
-          break;
-        case 'json':
-          result = JSON.parse(result);
-          break;
-        case 'jsonString':
-          // Stringify the jsonstring. This has the effect of double escaping the json, so that
-          // when we go to parse the final template to apply it to kube, it doesnt mistakenly
-          // turn our jsonString into actual json.
-          result = JSON.stringify(result);
-          // JSON.stringify adds quotes around the newly created json string. Kube forces us
-          // to wrap out curly braces in quotes so that it wont error on our templates. In order
-          // to avoid having 2 double quotes around the result, we need to remove the stringify
-          // quotes. slice(start of slice, end of slice)
-          result = result.slice(1, result.length - 1);
-          break;
-        case 'base64':
-          result = Buffer.from(result).toString('base64');
-          break;
+
+    const data = resource?.data;
+    if (kubeError === undefined) kubeError = 'no data returned from processEnvFrom. make sure your data exists in the correct location and is in the expected format.';
+
+    if (!data) {
+      if (!optional) {
+        throw new Error(`envFrom.genericMapRef.${apiVersion}.${kind}.${name}.data not found: ${kubeError}`);
+      } 
+
+      let msg = `envFrom.genericMapRef.${apiVersion}.${kind}.${name}.data not found: ${kubeError}`;
+      log.warn(msg);
+      this.updateRazeeLogs('warn', { controller: 'FetchEnvs', message: msg });
+    }
+
+    return {
+      ...conf,
+      data
+    };
+  }
+
+  async #genericKeyRef(conf) {
```

Comment:
I kind of feel like these could all be generic lookups under the hood

```javascript
secretKeyRef() {
  return this.genericKeyRef({ kind, apiVersion, key})
}
```

Reply:
Not sure if there is a performance penalty for the generic lookup. It looked like its pinging the API for some additional information to do this.

Reply:
i theory i dont think there would be too much performance penalty. it does have to lookup the krm, but that gets cached after the first lookup here https://github.com/razee-io/Kubernetes-util/blob/master/lib/kubeClass.js#L234, so shouldnt be horrible. 

i assume thats the api ping you are referring?

Reply:
theres def more overhead since secrets and configmaps we know the uri off the bat. by using generic, we have to do some sort of lookup to get the extra info in order to generate the kubeapi uri.

Reply:
Ya, thats what I was referring to the extra lookup.

Reply:
what if we combine the code like you suggest, but add an extra uri parameter in the function to short circuit the extra lookup?

Reply:
i think its a simple optimization that improves code readability and a perfomance improvement ( i dont know the cost of the lookup or if theres something im forgetting, so hard to say how much of an improvement ), i still dont think its a bad idea though